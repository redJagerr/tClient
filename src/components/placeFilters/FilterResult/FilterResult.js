import { useDispatch, useSelector } from 'react-redux';
import { useCallback, useEffect, useRef } from 'react';
import { useParams } from 'react-router-dom';
import { useSearchParams } from 'react-router-dom';
import { fetchPlaces } from 'redux/slices/filterResultSlice';
import {
	updateSort,
	toggleMobileVisible,
	toggleSwitch,
	updateAllFilters,
} from 'redux/slices/filtersSlice';
import {
	selectFilterResult,
	selectFilterResultTotal,
	selectFilters,
	selectFiltersSort,
} from 'redux/selectors/filtersSelectors';
import {
	selectCityName,
	selectSearchFilter,
} from 'redux/selectors/searchSelectors';
import PlaceCard from 'components/placeFilters/PlaceCard/PlaceCard';
import PlaceCardSkeleton from 'components/placeFilters/PlaceCard/PlaceCardSkeleton';
import PaginationNav from '../PaginationNav/PaginationNav';
import FiltersSelect from '../FiltersSelect/FiltersSelect';

import setUrlRequest from 'utils/setUrlRequest';
import declarationOfNumber from 'utils/declarationOfNumber';
import { BsSliders } from 'react-icons/bs';

const FilterResult = () => {
	const [searchParams, setSearchParams] = useSearchParams();
	const { cityId } = useParams();
	const dispatch = useDispatch();
	const initial = useRef(true);

	const totalPlaces = useSelector(selectFilterResultTotal);
	const name = useSelector(selectCityName);
	const searchFilter = useSelector(selectSearchFilter);
	const sort = useSelector(selectFiltersSort);
	const filters = useSelector(selectFilters);
	const { places, loadStatus } = useSelector(selectFilterResult);

	const toggleOfUrl = useCallback(() => {
		const sortBy = searchParams.has('sortBy')
			? searchParams.get('sortBy')
			: 'name';

		const orderBy = searchParams.has('orderBy')
			? searchParams.get('orderBy')
			: 'asc';

		const highRating = searchParams.has('rating');
		const localsChoice = searchParams.has('localsChoise');
		const page = searchParams.has('page') ? searchParams.get('page') : 1;

		if (sortBy !== 'name' || orderBy !== 'asc')
			dispatch(updateSort([sortBy, orderBy]));

		if (highRating) dispatch(toggleSwitch('highRating'));
		if (localsChoice) dispatch(toggleSwitch('localChoice'));

		dispatch(
			updateAllFilters({
				category: searchParams.getAll('category'),
				subcategory: searchParams.getAll('subcategory'),
				district: searchParams.getAll('district'),
				sort: [sortBy, orderBy],
				highRating,
				localsChoice,
				currentPage: Number(page),
			})
		);
	}, []);

	useEffect(() => {
		toggleOfUrl();
	}, []);

	useEffect(() => {
		if (initial.current) {
			initial.current = false;
			return;
		}
		const request = setUrlRequest(filters, searchFilter);
		dispatch(
			fetchPlaces({
				id: cityId,
				filters: request,
			})
		);
		setSearchParams(request);
	}, [filters, searchFilter]);

	const definePlacesOrder = e => {
		return e.target.value === 'likes' || e.target.value === 'rating'
			? 'desc'
			: 'asc';
	};

	const renderPlace = arr =>
		arr.map(({ id, ...props }) => (
			<PlaceCard key={id} cityId={cityId} placeId={id} {...props} />
		));

	const changeSort = e =>
		dispatch(updateSort([e.target.value, definePlacesOrder(e)]));

	const filteredPlaces = renderPlace(places);

	const isLoading = loadStatus === 'loading';
	const isError = loadStatus === 'error';

	const placesCount = `${name}: найден${
		places.length === 1 ? ' ' : 'о'
	} ${totalPlaces} ${declarationOfNumber(places.length, [
		'место',
		'места',
		'мест',
	])}`;

	return (
		<>
			<div className='places-filter__places filter-result'>
				<div className='filter-result__info'>
					<span className='filter-result__title places-filter__title'>
						{placesCount}
					</span>
					<div className='filter-result__buttons-container'>
						<FiltersSelect changeSort={changeSort} sort={sort[0]} />
						<button
							className='filter-result__button filter-result__button--filter'
							onClick={() => dispatch(toggleMobileVisible())}
						>
							<BsSliders size={24} fill='currentColor' />
						</button>
					</div>
				</div>
				{isError ? (
					<span>ОШИБКА! Не удалось связаться с сервером </span>
				) : (
					<ul className='filter-result__list'>
						{isLoading ? <PlaceCardSkeleton /> : filteredPlaces}
					</ul>
				)}
				<PaginationNav />
			</div>
		</>
	);
};
export default FilterResult;
