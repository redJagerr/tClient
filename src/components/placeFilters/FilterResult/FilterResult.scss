.filter-result {
	display: flex;
	flex-direction: column;
	min-width: 71%;
	min-height: toRem(300);

	&__info {
		display: flex;
		justify-content: space-between;
		margin-bottom: toRem(18);
	}
	&__sort {
		display: inline-block;
		font-size: toRem(11);
		font-weight: 500;
		color: var(--color-gray);
	}
	&__select {
		margin-left: toRem(5);
		font-size: inherit;
		font-weight: inherit;
		border: none;
		background: white;
		border-radius: 10px;
		&:active {
			border: none;
		}
	}
	&__option {
		font-weight: inherit;
		border-radius: 20px;
	}
	&__list {
		display: grid;
		grid-template-columns: repeat(2, 1fr);
		grid-auto-rows: minmax(toRem(232), auto);
		gap: toRem(20);
		margin-bottom: toRem(30);
		flex-shrink: 1;
	}
	&__buttons {
		display: flex;
		justify-content: space-between;
		margin-top: auto;
	}
	&__buttons-container {
		display: flex;
		gap: toRem(5);
		align-items: center;
		justify-content: space-between;
	}
	&__button {
		color: inherit;
		min-width: toRem(36);
		min-height: toRem(36);
		padding-top: toRem(5);
		background: none;
		border: solid 1px var(--color-gray);
		border-radius: 10px;
		&:hover {
			background-color: var(--primary-color);
			border: none;
			color: white;
		}
		&--active {
			background-color: var(--primary-color);
			border: none;
			color: white;
		}
		&--number {
			padding-top: 0;
			border: none;
		}
		&--filter {
			display: none;
			max-height: toRem(32);
		}
	}
	@include tablet {
		width: 100%;

		&__sort {
			font-size: 2.6vw;
		}
		&__info {
			flex-direction: column;
		}
		&__button {
			&--filter {
				display: inline-block;
			}
		}
	}
	@include mobile {
		&__list {
			grid-auto-rows: minmax(toRem(180), auto);
		}
	}
}
