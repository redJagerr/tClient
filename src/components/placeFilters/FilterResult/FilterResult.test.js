import FilterResult from './FilterResult';
import * as reduxHooks from 'react-redux';

import {
	selectFilterResult,
	selectFilterResultTotal,
	selectFilters,
	selectFiltersSort,
} from 'redux/selectors/filtersSelectors';
import {
	selectCityName,
	selectSearchFilter,
} from 'redux/selectors/searchSelectors';

import { renderWithRoutes } from 'tests/helpers/renderWithRoutes';

const mockState = {
	filterResult: {
		places: [],
		totalPlaces: 2,
		loadStatus: 'idle',
	},
	search: {
		searchFilter: '',
		cityInfo: {
			name: 'name',
		},
	},
	filters: {
		filters: {
			sort: ['name', 'desc'],
		},
	},
};
jest.mock('react-redux');
jest.mock('redux/selectors/searchSelectors');
jest.mock('redux/selectors/filtersSelectors');

describe('FilterResult component', () => {
	beforeEach(() => {
		const mockUseSelector = jest
			.spyOn(reduxHooks, 'useSelector')
			.mockImplementation(cb => cb(mockState));
		let mockUseDispatch = jest.spyOn(reduxHooks, 'useDispatch');
		const dispatch = jest.fn();
		mockUseDispatch.mockReturnValue(dispatch);
		selectFilterResult.mockReturnValue(mockState.filterResult);
		selectFilterResultTotal.mockReturnValue(mockState.filterResult.totalPlaces);
		selectFilters.mockReturnValue(mockState.filters);
		selectFiltersSort.mockReturnValue(mockState.filters.filters.sort);
	});
	it('should match with snapshot', () => {
		const component = renderWithRoutes(<FilterResult />);
		expect(component).toMatchSnapshot();
	});
});
