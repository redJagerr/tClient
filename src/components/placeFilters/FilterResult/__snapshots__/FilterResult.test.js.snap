// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterResult component should match with snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="places-filter__places filter-result"
      >
        <div
          class="filter-result__info"
        >
          <span
            class="filter-result__title places-filter__title"
          >
            undefined: найдено 2 мест
          </span>
          <div
            class="filter-result__buttons-container"
          >
            <label
              class="filter-result__sort"
              for="#"
            >
              Поиск по:
              <select
                class="filter-result__select"
                id="sort-by"
                name="sort-by"
              >
                <option
                  class="filter-result__option"
                  value="name"
                >
                  имени
                </option>
                <option
                  class="filter-result__option"
                  value="likes"
                >
                  лайкам
                </option>
              </select>
            </label>
            <button
              class="filter-result__button filter-result__button--filter"
            >
              <svg
                fill="currentColor"
                height="24"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 16 16"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11.5 2a1.5 1.5 0 1 0 0 3 1.5 1.5 0 0 0 0-3zM9.05 3a2.5 2.5 0 0 1 4.9 0H16v1h-2.05a2.5 2.5 0 0 1-4.9 0H0V3h9.05zM4.5 7a1.5 1.5 0 1 0 0 3 1.5 1.5 0 0 0 0-3zM2.05 8a2.5 2.5 0 0 1 4.9 0H16v1H6.95a2.5 2.5 0 0 1-4.9 0H0V8h2.05zm9.45 4a1.5 1.5 0 1 0 0 3 1.5 1.5 0 0 0 0-3zm-2.45 1a2.5 2.5 0 0 1 4.9 0H16v1h-2.05a2.5 2.5 0 0 1-4.9 0H0v-1h9.05z"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
        </div>
        <ul
          class="filter-result__list"
        />
        <div
          class="filter-result__buttons"
        >
          <button
            class="button filter-result__button"
          >
            <svg
              fill="currentColor"
              height="20"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 8a.5.5 0 0 1-.5.5H5.707l2.147 2.146a.5.5 0 0 1-.708.708l-3-3a.5.5 0 0 1 0-.708l3-3a.5.5 0 1 1 .708.708L5.707 7.5H11.5a.5.5 0 0 1 .5.5z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
          <div
            class="filter-result__buttons-container"
          />
          <button
            class="button filter-result__button "
          >
            <svg
              fill="currentColor"
              height="20"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4 8a.5.5 0 0 1 .5-.5h5.793L8.146 5.354a.5.5 0 1 1 .708-.708l3 3a.5.5 0 0 1 0 .708l-3 3a.5.5 0 0 1-.708-.708L10.293 8.5H4.5A.5.5 0 0 1 4 8z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="places-filter__places filter-result"
    >
      <div
        class="filter-result__info"
      >
        <span
          class="filter-result__title places-filter__title"
        >
          undefined: найдено 2 мест
        </span>
        <div
          class="filter-result__buttons-container"
        >
          <label
            class="filter-result__sort"
            for="#"
          >
            Поиск по:
            <select
              class="filter-result__select"
              id="sort-by"
              name="sort-by"
            >
              <option
                class="filter-result__option"
                value="name"
              >
                имени
              </option>
              <option
                class="filter-result__option"
                value="likes"
              >
                лайкам
              </option>
            </select>
          </label>
          <button
            class="filter-result__button filter-result__button--filter"
          >
            <svg
              fill="currentColor"
              height="24"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11.5 2a1.5 1.5 0 1 0 0 3 1.5 1.5 0 0 0 0-3zM9.05 3a2.5 2.5 0 0 1 4.9 0H16v1h-2.05a2.5 2.5 0 0 1-4.9 0H0V3h9.05zM4.5 7a1.5 1.5 0 1 0 0 3 1.5 1.5 0 0 0 0-3zM2.05 8a2.5 2.5 0 0 1 4.9 0H16v1H6.95a2.5 2.5 0 0 1-4.9 0H0V8h2.05zm9.45 4a1.5 1.5 0 1 0 0 3 1.5 1.5 0 0 0 0-3zm-2.45 1a2.5 2.5 0 0 1 4.9 0H16v1h-2.05a2.5 2.5 0 0 1-4.9 0H0v-1h9.05z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </div>
      </div>
      <ul
        class="filter-result__list"
      />
      <div
        class="filter-result__buttons"
      >
        <button
          class="button filter-result__button"
        >
          <svg
            fill="currentColor"
            height="20"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 8a.5.5 0 0 1-.5.5H5.707l2.147 2.146a.5.5 0 0 1-.708.708l-3-3a.5.5 0 0 1 0-.708l3-3a.5.5 0 1 1 .708.708L5.707 7.5H11.5a.5.5 0 0 1 .5.5z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
        <div
          class="filter-result__buttons-container"
        />
        <button
          class="button filter-result__button "
        >
          <svg
            fill="currentColor"
            height="20"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 8a.5.5 0 0 1 .5-.5h5.793L8.146 5.354a.5.5 0 1 1 .708-.708l3 3a.5.5 0 0 1 0 .708l-3 3a.5.5 0 0 1-.708-.708L10.293 8.5H4.5A.5.5 0 0 1 4 8z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
